src/main/java/com/spotify/dns/ServiceResolvingChangeNotifier.java

```
@@ -0,0 +1,104 @@
+
+  private final DnsSrvResolver resolver;
+  private final String fqdn;
+  private final Function<LookupResult, T> resultTransformer;
+
+  private volatile Set<T> records = Collections.emptySet();
+
+  private volatile boolean run = true;
+
+  /**
+   * Create a {@link ChangeNotifier} that tracks changes from a {@link DnsSrvResolver}.
+   *
+   * The list of {@link LookupResult}s will be transformed using the provided function
+   * and put into a set. The set will then be compared to the previous set and if a
+   * change is detected, the notifier will fire.
+   *
+   * @param resolver            The resolver to use.
+   * @param fqdn                The name to lookup SRV records for
+   * @param resultTransformer   The transform function
+   */
+  ServiceResolvingChangeNotifier(final DnsSrvResolver resolver,
+                                 final String fqdn,
+                                 final Function<LookupResult, T> resultTransformer) {
+
+    this.resolver = checkNotNull(resolver);
+    this.fqdn = checkNotNull(fqdn, "fqdn");
+    this.resultTransformer = checkNotNull(resultTransformer, "resultTransformer");
+  }
+
+  @Override
+  protected void closeImplementation() {
+    run = false;
+  }
+
+  @Override
+  public Set<T> current() {
+    return Collections.unmodifiableSet(records);
+  }
+
+  @Override
+  public void run() {
+    if (!run) {
+      return;
+    }
+
+    try {
+      List<LookupResult> nodes = resolver.resolve(fqdn);
+      Set<T> currentRecords = Sets.newHashSet();
+      for (LookupResult node : nodes) {
+        currentRecords.add(resultTransformer.apply(node));
+      }
+
+      if (!currentRecords.equals(records)) {
+        final ChangeNotification<T> changeNotification =
+            newChangeNotification(currentRecords, records);
+        records = currentRecords;
+
+        fireRecordsUpdated(changeNotification);
+      }
+    } catch (Exception e) {
+      log.error(e.getMessage(), e);
```

maybe propagate the error to listeners somehow?